### ISSUE TITLE
[scan] stop_decision's incumbent is the max of noisy observations, so one lucky assay stops the campaign — the exact failure direction the module was built to avoid

### ISSUE THREAD

**Issue by dcbaugher** [enginbio/engin-suite#250]
**Lens:** Methodology (rotation idx 3) · **Scanned:** 2026-08-20

## What

`stop_decision(mean, sd, best, ...)` treats `best` as the incumbent to beat, and every producer of that value in this repo passes the maximum of the *noisy observed* titers — `float(np.max(y))` at `packages/engin-core/src/engin_core/cli.py:61`, `float(y_obs.max())` at `packages/engin-core/benchmarks/benchmark.py:208,344`, `float(y_obs.max())` at `packages/engin-core/examples/run_demo.py:82`. The max of n noisy draws is an upward-biased estimate of the best achievable titer, and the bias is largest exactly when it hurts: one assay that reads high by two or three sigma becomes a permanent threshold that nothing can beat, and `headroom` returns ~0 for every candidate forever after.

This matters far more in `stopping.py` than in `recommend.py`. In the EI recommender `best` is a ranking offset and the ranking is robust to it; in the stopping rule it is an absolute threshold on the objective, so the bias passes through undamped.

## Why it matters

`stopping.py:27` states the module's own design constraint: *"an overconfident posterior thinks there is less left to find than there is, so it **stops early**"*, and the whole conformal-rescaling apparatus (`_calibrated_scale`) exists to widen the predictive so the rule does not stop early. A noise-inflated incumbent stops it early anyway, through the other argument, and by a larger margin than the conformal widening corrects.

Measured on this repo's own simulator (24-point DoE + one round of 8 from `recommend_batch`, `add_noise(rel=0.05, abs_=0.4)` copied verbatim from `benchmarks/benchmark.py:164`, `epsilon=2.0`, `delta=0.05`, `q` from a real split-conformal calibration, 4000-point pool, seeds 0–11):

```
 seed  best_obs  y_true at that pt   noise    max y_true | stop(obs)  stop(post-mean)
    3    126.51             111.76  +14.75        111.76 | True       False   <-- stalled
    4    119.15             111.76   +7.39        111.76 | True       False   <-- stalled
    5    118.30             111.76   +6.54        111.76 | True       False   <-- stalled
    6    114.59             108.16   +6.43        108.16 | True       False   <-- stalled
    8    112.62             111.76   +0.86        111.76 | True       False   <-- stalled

campaigns stopped after ONE round on an incumbent no process point can reach: 5/12
```

Five of twelve seeds declare the campaign finished after a single round, on a threshold above the global maximum of the true response surface. Seed 3's incumbent is 14.75 g/L of pure measurement noise. Substituting the posterior mean at the observed designs — `float(gp.predict(gp.X, include_noise=False)[0].max())` — stops in **zero** of the twelve.

A process lead who is told "stop, there is nothing left to find" after one round, because one shake flask read high, has been given a worse answer than no stopping rule at all.

## Evidence

- `packages/engin-core/src/engin_core/stopping.py:27` — "so it **stops early**", the stated dangerous direction.
- `packages/engin-core/src/engin_core/stopping.py:97-117` — `headroom` computes `P(f(x) > best + epsilon)`; `best` enters as an absolute threshold.
- `packages/engin-core/src/engin_core/cli.py:61`, `benchmarks/benchmark.py:208,344`, `examples/run_demo.py:82` — every incumbent in the repo is `max(noisy y)`.
- `packages/engin-core/src/engin_core/gp.py:120-122` — `fit_gp` fits `WhiteKernel(0.1, (1e-4, 1.0))`, i.e. the model explicitly believes observations are noisy. The incumbent ignores the noise model the same function just fitted.
- Smith & Winkler, *The Optimizer's Curse: Skepticism and Postdecision Surprise in Decision Analysis*, Management Science 52(3):311-322 (2006), https://doi.org/10.1287/mnsc.1050.0451 — the general result that selecting the max of noisy estimates yields an optimistically biased estimate of the selected item's true value.
- Gramacy & Lee, *Optimization Under Unknown Constraints*, https://arxiv.org/abs/1004.4027 — "Another reasonable choice that we promote in this paper is f_min = min ẑ_N(x), the minimum of the (posterior) mean predictive surface." The canonical statement of the fix.
- Wang, Wang, Ng & Petra, *Bayesian Optimization with Expected Improvement: No Regret and the Choice of Incumbent*, arXiv:2508.15674, https://arxiv.org/html/2508.15674v1 — names the three options (best-posterior-mean, best-*sampled*-posterior-mean, best-observation) and proves sublinear cumulative regret for the first two while the best-observation incumbent is not established as no-regret.
- BayBE (Merck) — the DoE library `D9` names as a compose target — takes the best-sampled-posterior-mean route: `_set_best_f` sets `best_f = self._posterior_mean_comp.max().item()`, the posterior mean of the transformed training targets. https://github.com/emdgroup/baybe/blob/main/baybe/acquisition/_builder.py
- BoTorch documents `best_f` as "the best function value observed so far **(assumed noiseless)**" and ships `qLogNoisyExpectedImprovement`, which "does not assume a `best_f` is known (which would require noiseless observations)". https://github.com/pytorch/botorch/blob/main/botorch/acquisition/analytic.py

## What it would take

Smallest version that settles it: add an `incumbent=` argument to `stop_decision` and `headroom`, default it to a posterior-mean incumbent computed from `gp`, and re-run the table above. That is a few lines plus a test asserting that a single outlier observation does not flip `stop` to `True`. If the maintainer prefers not to change the signature, the same result comes from documenting that callers must pass a de-noised incumbent and fixing the four call sites — but a library whose contract is honest uncertainty should not leave the correct usage to the caller's memory (`D14`).

The larger version — replacing `best` with Wilson's within-sample-path regret — is #18's territory and is not proposed here.

## Decisions touched

`D9` — neutral-to-implements. The fix is one line of existing GP output, not a new dependency; but if the eventual answer is "use BoTorch/BayBE's noisy acquisition machinery", that is `D9` working as intended.
`D12` — relevant: the benchmarks report campaign outcomes computed with this incumbent, so any published stop-round number inherits the bias.
`D14` — the signature change is API surface; keep it additive with a default.

## Why this might be wrong

**The strongest objection, and it nearly killed this issue: the same defect is almost harmless in the EI recommender, and I originally filed it there.** Measured over 8 seeds at n=32, swapping to the posterior-mean incumbent changed the top-8 of the 4000-point EI pool by 0.25 picks out of 8 (7.75/8 overlap), the incumbent shifted by 0.81 g/L on an 87.8 g/L scale, and the fraction of the pool with EI underflowing to exactly 0.0 was 1.6% either way. So the literature's framing — "the noisy-max incumbent is brittle" — does *not* translate into a measurable recommender defect here, and the issue would have been wrong if filed that way. What survives is specifically the stopping rule, where `best` is a threshold rather than an offset.

Second objection: the 5/12 figure is on the bundled simulator with a synthetic noise model, and `add_noise`'s 5% relative term may be pessimistic for a well-run titer assay. If real replicate CV is 1-2%, the stall rate falls. It does not fall to zero — the mechanism is one-sided regardless of magnitude — but the headline number is simulator-dependent and should be read as an existence proof, not a rate.

Third: `stop_decision` has no call site in the CLI, docs, examples or benchmarks today — only tests. So nothing currently ships this behaviour to a user, and the fix is cheap precisely because it is early. That is an argument about urgency, not about correctness.

Fourth, and worth stating because it cuts against the framing: the bioprocess field's own current review — Siska et al., *A Guide to Bayesian Optimization in Bioprocess Engineering*, Biotechnol. Bioeng. 123(4):805-830 (2026), https://arxiv.org/html/2508.10642v1 — presents EI with "f(x*) is the best observed value so far" with no caveat, and Optuna's `GPSampler` still uses the raw observed max. Engin is not an outlier within its field. It is an outlier against the ML literature and against Ax, BoFire, BayBE and Trieste, all of which use a posterior-based incumbent.

**Comment by dcbaugher** [comment]
Taking this together with the sibling in `stopping.py` — they are the same function and the same decision, and fixing one alone leaves the rule half-right. Verified both claims against `origin/main` before claiming: `stopping.py:146` is the product, and `headroom` takes `best` as an absolute threshold.

Plan is the smallest version each issue proposes, and I will reproduce the reported numbers *before* changing anything so the fix is measured against a reproduction rather than against the write-up:

- **#250** — additive `incumbent=` on `stop_decision`/`headroom`, defaulting to a posterior-mean incumbent, keeping the signature backward-compatible (`D14`).
- **#251** — report `max_i p_i` and stop when that falls below `delta`, which is pool-size invariant and converges correctly under refinement.

Not doing the larger versions: joint posterior sampling per Wilson, and regret-based stopping, both of which the issues put in #18. If reproduction disagrees with either write-up I will report that rather than fix past it.

**Comment by dcbaugher** [comment]
**Reproduction note, since two things in the write-up did not survive it.** Details in #254.

**1. `stop_decision` has no production callers.** This issue cites `cli.py:61`, `benchmark.py:208,344` and `run_demo.py:82` as the producers of a noisy incumbent. Those are `recommend_batch(best_y=...)` — the EI recommender, not the stopping rule. Grepping `stop_decision` and `engin_core.stopping` across `*.py`, `*.md`, `*.ipynb`, `*.yaml` and `*.toml` finds it only in `stopping.py` and `tests/test_stopping.py`, and it is not exported from `engin_core/__init__`. So both defects are **latent** — real in the contract, reached by nothing today. That does not make them not worth fixing (the module is under the api-stability guarantee and #18 will put a loop around it), but the severity framing should be read accordingly, the same way #224 read its own.

**2. "Substituting the posterior mean ... stops in zero of the twelve" did not reproduce.** In my harness a posterior-mean incumbent stops **8 of 12**. What does reproduce is the differential: **3 of 12** campaigns stop only because the incumbent is inflated, and the observed maximum sits above the *true* maximum of the response surface in **10 of 12**, by a median of +2.4 g/L and as much as +9.9 — your +9.9 figure lands almost exactly.

The mechanism is confirmed and the fix is unchanged. The docstrings carry my measured numbers rather than the reported ones, because the two disagree and only one of them was run against this checkout.
